marker_lints/src/not_using_has_span_trait.rs

```diff
@@ -0,0 +1,44 @@
+use marker_api::prelude::*;
+
+marker_api::declare_lint! {
+    /// ### What it does
+    /// Suggests using `impl HasSpan` for functions that need to take `Span` as
+    /// a parameter to make them more ergonomic.
+    NOT_USING_HAS_SPAN_TRAIT,
+    Warn,
+}
+
+pub(crate) fn check_item<'ast>(cx: &'ast MarkerContext<'ast>, item: ItemKind<'ast>) {
```

Comment:
I'm struggling with code organization here. Since there is a single lint pass for the entire crate, then if I want to define unrelated lints in separate modules I can't write their code inside of a trait impl. Writing code inside of a trait impl is cool because I can leverage rust-analyzer to paste the stub trait impl and let me fill it in. With the free function I had to copy the signature from the trait's declaration manually. I wish we could improve this situation.

Reply:
Yeah, I'm currently experimenting a bit in https://github.com/rust-marker/marker-example-lints. Maybe having something like the inventory crate, that you talked about previously could be a good idea :thinking:

Reply:
Yeah, I already forgot about `inventory`. I'm thinking what the right API should be though, it's hard to to come up with something both succinct but flexible.

Reply:
I see that `clippy` uses a single lint pass per lint.
For example `MultiAssignments` is defined in a separate file with its own lint pass ([code](https://github.com/rust-lang/rust-clippy/blob/33f49f33d6f8cec9a5772bc7f3a4e5f1de207b30/clippy_lints/src/multi_assignments.rs#L32)).

And then that lint pass is registered manually ([code](https://github.com/rust-lang/rust-clippy/blob/33f49f33d6f8cec9a5772bc7f3a4e5f1de207b30/clippy_lints/src/lib.rs#L1010)).

That second part of manual registration can be done by `inventory` though.

But let's not forget that lints should be configurable, like [this one](https://github.com/rust-lang/rust-clippy/blob/33f49f33d6f8cec9a5772bc7f3a4e5f1de207b30/clippy_lints/src/enum_variants.rs#L107-L112).

Reply:
Clippy usually has one lint pass per lint crate, but we also combine them sometimes. This can become messy (See [`clippy_lints::methods`](https://github.com/rust-lang/rust-clippy/blob/master/clippy_lints/src/methods/mod.rs) as an extreme example), but allows lints to share common logic.

We could do something similar, where we have a lint pass trait implemented on several types. The main `LintPass` of the crate would hold `Vec<Box<&dyn LintPass>>`, to distribute the control flow inside the crate. The `dyn LintPass` wouldn't have access to the `'ast` lifetime, but I think we want to discourage users from storing any `'ast` objects in the structs anyways

Reply:
Aren't clippy lints all in one crate? I've shown the example where they register one lint pass per lint module.

That extreme example is really extreme. It's hard to tell which logic they share in common, and wether it's even worth it. I suppose it's an attempt to save some performance on looking up method expressions in code. I.e. the difference between:
```rust
if let Expr::Method(method) = expr {
	method_lint_1(method);
	method_lint_2(method);
}
```
```rust
if let Expr::Method(method) = expr {
	method_lint_1(method);
}

if let Expr::Method(method) = expr {
	method_lint_2(method);
}
```
and the first one wins because it doesn't have a repeated check?
That makes sense, but I'm not sure such optimization is worth it, given the messy code organization, but I suppose we shouldn't have an API that disallows doing such context sharing.

Maybe there could be several variations of `declare_lint!` macro. One for the simple case where the lint with the lint pass implementation is directly registered in `inventory`, and the second one more advanced that registers only the lint in the inventory, but the implementation of the lint pass is expected to be used manually inside of other lint pass implementation and thus not registered.